### PR TITLE
Optimize mocking further by retaining mocks across tests (alt 2)

### PR DESCRIPTION
**Pull request by tomas-abrahamsson**
This is an alternative solution to the #20, as discussed.

Just like the #20, it takes the optimizations in #18 a bit further by retaining mocked modules between tests in a `?WITH_MOCKED_SETUP`. That is, between such tests, the mocked modules are not restored to original, but instead the modules are left in a mocked state, the next test will need them mocked. But the state machine steps to do that is a bit different from how it is done in the #20, and the strategy is this:

* Use an [eunit fixture ](https://erlang.org/doc/apps/eunit/chapter.html#Fixtures) to set up mocking and unload mocked modules. The fixure will start and end a session.
* Inside each session, we can have several session elements, each corresponds to an eunit test.
* The session elements handles call tracing and actions.

I had difficulties coming up with a good name for parts of a session (and equally difficult for the opposite of something that encompasses several sessions if going that direction instead) so for now I've used the term 'session element'. Any suggestions are of course welcome :) 

Just like the #20, this PR, too, introduces a slight backwards incompatible change: If unit test code in the setup of a ?WITH_MOCKED_SETUP is calling a mocked module, the call gets routed over to the renamed module, the `mod^`.  This usually works, but there can be corner cases if the renamed module uses the `?MODULE` in creative ways.

No tests in larger projects have yet been run. I expect the speedup to be on par with the #20, but haven't measured. I also have not tried it interactively very much to see if there may be any issues with the use of eunit fixtures. I think it may be wise to do that, just to verify that this approach never leaves sessions or elements hanging under any circumstances.

### CODE REVIEW

File: src/mockgyver.erl [hunk 1]
```diff
@@ -414,10 +414,10 @@
 %%%===================================================================
 
 %% @private
-exec(MockMFAs, WatchMFAs, Fun) ->
+exec(Context, Fun) ->
```
Comment (klajo) [inline src/mockgyver.erl:425]: 1. Thinking it might be good to -spec this, now that the name of the variables don't hint at what it is anymore, and/or
2. Is Context the best name here?

I'm thinking it's not the context in which we're doing this. We're mocking and watching MFAs in some context.  Something with MFA?

File: src/mockgyver_xform.erl [hunk 2]
```diff
@@ -177,21 +177,26 @@ is_mock_expr(tuple, Form) ->
     case erl_syntax:tuple_elements(Form) of
         [H | T] ->
             case erl_syntax:is_atom(H, '$mock') of
-                true  -> {true, analyze_mock_form(T)};
-                false -> false
+                true  ->
+                    [TypeTree | Rest] = T,
+                    Type = erl_syntax:atom_value(TypeTree),
+                    Location = erl_syntax:concrete(lists:last(T)),
```
Comment (klajo) [inline src/mockgyver_xform.erl:207]: Minor, use lists:last(Rest) and swap 183 and 184 (just to keep them in the other they come?

File: include/mockgyver.hrl [hunk 3]
```diff
@@ -8,19 +8,31 @@
 %% run tests with a mock
```
Comment (klajo) [inline include/mockgyver.hrl:8]: Didn't find how to add this on line 4: dependency no longer needed (unless we refactor the fixture into eunit_addons)?

File: include/mockgyver.hrl [hunk 4]
```diff
@@ -30,9 +30,14 @@
...
 
+-define(WRAP(Type),
+        {'$mock', Type, {?FILE, ?LINE}}).
+
 -define(WRAP(Type, Expr),
         {'$mock', Type, Expr, {?FILE, ?LINE}}).
 
+-define(MOCK_CONTEXT, ?WRAP(m_context)).
+
 -define(MOCK(Expr), ?WRAP(m_init, (Expr))).
```
Comment (klajo) [inline include/mockgyver.hrl:53]: I think this could now be implemented in terms of mockgyver:exec(?MOCK_CONTEXT, Expr). Or, if there is a reason for keeping it, to rename the m_init to something that speaks of one-shot mocks (or something to that effect).

**Review by klajo — CHANGES_REQUESTED** [review]
Thanks! Added comments.
As for the naming of the sessions: would it be a [herd](https://en.wikipedia.org/wiki/List_of_animal_names#Usage_of_collective_nouns) of sessions? Or are they more like insects so it should be a swarm of sessions? ;-)

More testing sounds good.

File: src/mockgyver.erl (hunk 1 above)
Comment (tomas-abrahamsson) [reply src/mockgyver.erl:425]: I agree it would be good to spec this. I was first thinking specing should perhaps be a separate pr since there are (almost) no specs and several @specs which render doc warnings now with Erlang 24. But then I realized there is actually one -spec, so ok. Will do. Due to the name-depletion (below), the also-added type will for now be context().

Naming: Well... I seem to have run out of names at the moment. Can't find any around here. Will return to this when the stores are open again. I'll (force) push the branch with the rest of the changes for now.

File: src/mockgyver_xform.erl (hunk 2 above)
Comment (tomas-abrahamsson) [reply src/mockgyver_xform.erl:207]: will do

File: include/mockgyver.hrl (hunk 4 above)
Comment (tomas-abrahamsson) [reply include/mockgyver.hrl:53]: Good point. I added a commit that implements it as mockgyver:exec(?MOCK_CONTEXT, Expr), and removes m_init from the parse transform.

**Comment by tomas-abrahamsson** [comment]
> ... swarm of sessions? ;-)

Would that be crowd testing?

> More testing sounds good.

An update is that some larger-project testing succeeded, but I haven't yet come around to do any interactive testing.

File: src/mockgyver.erl (hunk 1 above)
Comment (tomas-abrahamsson) [reply src/mockgyver.erl:428]: Oh, apologies for accidentally pinging! It was unintentional.

Comment (klajo) [reply src/mockgyver.erl:428]: :-)

I've pushed a side-branch called specify with type specs added to exported functions and prominent data types.

File: include/mockgyver.hrl (hunk 3 above)
Comment (klajo) [reply include/mockgyver.hrl:8]: Good point. Let's leave it in.

**Comment by klajo** [comment]
> > ... swarm of sessions? ;-)
> 
> Would that be crowd testing?

:-)

> > More testing sounds good.
> 
> An update is that some larger-project testing succeeded, but I haven't yet come around to do any interactive testing.

Ok, so far so good.

File: src/mockgyver_xform.erl (hunk 2 above)
Comment (klajo) [reply src/mockgyver_xform.erl]: Perfect, thanks!

File: include/mockgyver.hrl (hunk 4 above)
Comment (klajo) [reply include/mockgyver.hrl]: Great, thanks!

File: src/mockgyver.erl (hunk 1 above)
Comment (tomas-abrahamsson) [reply src/mockgyver.erl]: It is now session params instead of context

**Comment by tomas-abrahamsson** [comment]
An update: I have run this for a while now, and have seen only one oddity. I have been running it both in an interactive inferior erlang session and repeatedly on the command line. I have also tried it in a couple of larger projects. I have not yet verified that this pr has the same speed-up as the #20.

The one oddity that I've seen is with a rebar2-based build system. The problem was that it did not rebuild eunit test beams when I switched from mockgyver's master to this PR. The result was an `undef` error when it was still calling `mockgyver:exec/3` which had been replaced by `mockgyver:exec/2`. However, eunit appears to depth limit the printout, and when the mocks are many, and if one is not the developer of this PR, it can probably be a bit tricky to figure out what is going on. It was initially non-obvious even for the PR author :) (The reverse is probably also possible, but downgrading mockgyver ought to be less common.) Anyways, cleaning is an easy way to get situation under control again. (I have not checked the same with a rebar3 based project where mockgyver is a dependency or _checkout. but I'd expect that it probably works better.)

One thing to consider, though, is perhaps whether to retain (for a while perhaps) a `mockgyver:exec/3` that just errors out with a reason or printout to rebuild or clean. Or alternatively, to consider retaining `exec/3` as is it looks on master, but then it might instead be necessary to split the `?MOCK_SESSION_PARAMS` macro into two.

**Comment by klajo** [comment]
Interesting that you say that, I noticed the exact same problem but got sidetracked and forgot to mention it to you.

Good to hear that there have been almost no problems. 

The option on retaining the exec/3 would mean that existing already compiled code would continue to work, but not benefit from the optimizations I assume. That should work. On the other hand the code is neater like it is now, isn't it? Users first, perhaps. 

Otherwise, when we conclude that part we're ready to go, right?

**Comment by tomas-abrahamsson** [comment]
I'll see if I can do something reasonable.
Yes, otherwise, I think it is ready to go.

**Comment by tomas-abrahamsson** [comment]
I think that the CI error for Erlang 24 will get fixed by #22

**Comment by klajo** [comment]
The exec/3 solution looks clean.

I merged #22 to master if you want to rebase this on top of that.

**Comment by tomas-abrahamsson** [comment]
> I merged #22 to master if you want to rebase this on top of that.

Done!

I've verified that speed-up is on par with #20. I realized this PR was still in "work in progress" status, so I hit the "Ready for review" button. A little late maybe :)

File: src/mockgyver.erl (hunk 1 above)
Comment (tomas-abrahamsson) [reply src/mockgyver.erl]: Guess we've settled on session params instead of context, so I'll try to mark this item as resolved. (Guess it is still possible to re-open if there's anything more)

**Comment by klajo** [comment]
Thanks! Now on master.
